homeassistant/components/aosmith/__init__.py

```diff
@@ -11,33 +9,37 @@
 from homeassistant.helpers import aiohttp_client
 
 from .const import DOMAIN
-from .coordinator import AOSmithCoordinator
+from .coordinator import AOSmithEnergyCoordinator, AOSmithStatusCoordinator
+from .models import AOSmithData, build_device_details
 
 PLATFORMS: list[Platform] = [Platform.SENSOR, Platform.WATER_HEATER]
 
 
-@dataclass
-class AOSmithData:
-    """Data for the A. O. Smith integration."""
-
-    coordinator: AOSmithCoordinator
-    client: AOSmithAPIClient
-
-
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up A. O. Smith from a config entry."""
     email = entry.data[CONF_EMAIL]
     password = entry.data[CONF_PASSWORD]
 
     session = aiohttp_client.async_get_clientsession(hass)
     client = AOSmithAPIClient(email, password, session)
-    coordinator = AOSmithCoordinator(hass, client)
 
-    # Fetch initial data so we have data when entities subscribe
-    await coordinator.async_config_entry_first_refresh()
+    status_coordinator = AOSmithStatusCoordinator(hass, client)
+    await status_coordinator.async_config_entry_first_refresh()
+
+    device_details_list = [
+        build_device_details(device) for device in status_coordinator.data.values()
+    ]
```

Comment:
Hmm, I don't really feel the whole abstraction to have a separate object for device details. Can we maybe just pass on the status_coordinator.keys() to the energy coordinator and call it a day? The in the sensor platform you can just get the data from the status coordinator and pass that into the new entities.

Reply:
The status coordinator data is just a `dict[str, dict[str, Any]]`. If you don't mind, I'd prefer to keep it this way because I really think it makes the code cleaner to have a strongly typed dataclass rather than just passing a dict around everywhere. I do agree with the first part of your suggestion though, about passing the status coordinator keys to the energy coordinator, and I've made that change.